File: flutter/lib/data/contact_repository.dart
import '../models/contact.dart';
import 'mock_contacts.dart';

/// Read access to the contact list.
///
/// The interface exists so the UI never depends on where contacts come from.
/// Week 4 ships the in-memory implementation below; swapping in an HTTP or
/// SQLite backed version later is a one-line change in `main.dart`.
abstract interface class ContactRepository {
  Future<List<Contact>> fetchContacts();
}

/// In-memory implementation backed by [kMockContacts].
class MockContactRepository implements ContactRepository {
  MockContactRepository({List<Contact>? seed, this.latency = Duration.zero})
      : _contacts = List<Contact>.unmodifiable(seed ?? kMockContacts);

  final List<Contact> _contacts;

  /// Artificial delay, so loading states can be demonstrated. Tests pass
  /// [Duration.zero] and therefore never need to wait on a real timer.
  final Duration latency;

  @override
  Future<List<Contact>> fetchContacts() async {
    if (latency > Duration.zero) {
      await Future<void>.delayed(latency);
    }
    return _contacts;
  }
}

File: mobile/src/data/contactRepository.ts
import type { Contact } from '../models/contact';
import { mockContacts } from './mockContacts';

/**
 * Read access to the contact list.
 *
 * The interface exists so the UI never depends on where contacts come from.
 * Week 5 ships the in-memory implementation below; swapping in an HTTP or
 * SQLite backed version later is a one-line change in `App.tsx`.
 */
export interface ContactRepository {
  fetchContacts(): Promise<Contact[]>;
}

/** In-memory implementation backed by the prototype's roster. */
export function createMockContactRepository(seed: Contact[] = mockContacts): ContactRepository {
  return {
    async fetchContacts() {
      return seed;
    },
  };
}

/**
 * A repository whose first `failures` loads fail, for exercising the error path
 * — and the recovery from it — through the real screen.
 */
export function createFlakyContactRepository(
  failures = 1,
  seed: Contact[] = mockContacts,
): ContactRepository {
  let attempts = 0;
  return {
    async fetchContacts() {
      attempts += 1;
      if (attempts <= failures) throw new Error('offline');
      return seed;
    },
  };
}
